pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -522,63 +501,138 @@ func (p *Pod) constructGroupPodSets(podsInGroup corev1.PodList) ([]kueue.PodSet,
+	if err != nil {
+		return err
+	}
+	originalQueue := jobframework.QueueName(p)
 
-	var (
-		podSets []kueue.PodSet
-	)
+	if p.countActivePods() < groupTotalCount {
+		errMsg := fmt.Sprintf("'%s' group total count is less than the actual number of pods in the cluster", p.groupName())
+		r.Eventf(p.Object(), corev1.EventTypeWarning, "ErrWorkloadCompose", errMsg)
+		return jobframework.UnretryableError(errMsg)
+	}
 
-	if !p.isGroup {
-		podSets = p.PodSets()
-	} else {
-		groupTotalCount, err := p.groupTotalCount()
-		if err != nil {
-			return nil, err
+	for _, podInGroup := range p.list.Items {
+		// Skip failed pods
+		if podInGroup.Status.Phase == corev1.PodFailed {
+			continue
 		}
 
-		podCount := 0
-		for i := range p.list.Items {
-			if p.list.Items[i].Status.Phase != corev1.PodFailed {
-				podCount++
-			}
+		if podInGroupQueue := jobframework.QueueNameForObject(&podInGroup); podInGroupQueue != originalQueue {
+			return jobframework.UnretryableError(fmt.Sprintf("pods '%s' and '%s' has different queue names: %s!=%s",
+				p.pod.GetName(), podInGroup.GetName(),
+				originalQueue, podInGroupQueue))
 		}
 
-		if podCount != groupTotalCount {
-			errMsg := fmt.Sprintf("'%s' group total count is different from the actual number of pods in the cluster", p.groupName())
-			r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", errMsg)
-			return nil, jobframework.UnretryableError(errMsg)
+		tc, err := strconv.Atoi(podInGroup.GetAnnotations()[GroupTotalCountAnnotation])
+		if err != nil {
+			return fmt.Errorf("failed to extract '%s' annotation from the pod '%s': %w",
+				GroupTotalCountAnnotation,
+				podInGroup.GetName(),
+				err)
+		}
+		if tc != groupTotalCount {
+			return jobframework.UnretryableError(fmt.Sprintf("pods '%s' and '%s' has different '%s' values: %d!=%d",
+				p.pod.GetName(), podInGroup.GetName(),
+				GroupTotalCountAnnotation,
+				groupTotalCount, tc))
 		}
+	}
 
-		podSets, err = p.constructGroupPodSets(p.list)
-		if err != nil {
-			if jobframework.IsUnretryableError(err) {
-				r.Eventf(object, corev1.EventTypeWarning, "ErrWorkloadCompose", err.Error())
+	return nil
+}
+
+// countActivePods returns a count of non-failed pods in the group
+func (p *Pod) countActivePods() int {
+	if p.activePodsCount != nil {
+		return *p.activePodsCount
+	}
+
+	// Filter all failed pods
+	activePodsCount := 0
+	for _, pod := range p.list.Items {
+		if pod.Status.Phase != corev1.PodFailed {
+			activePodsCount++
+		}
+	}
+
+	p.activePodsCount = &activePodsCount
+	return activePodsCount
+}
+
+// cleanupExcessPods will delete and finalize pods created last if the number of
+// activePods is greater than the group-total-count annotation value.
+func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	groupTotalCount, err := p.groupTotalCount()
+	if err != nil {
+		return err
+	}
+
+	extraPods := p.countActivePods() - groupTotalCount
+
+	for i := extraPods; i > 0; i-- {
```

Comment:
Better to obtain a list of non-terminated Pods and sort.

Reply:
Sure, I can do that. 

Just keep in mind, that I will have to search in the `p.list` for the deleted excess pods here:
https://github.com/kubernetes-sigs/kueue/blob/a2e0cf4ab5ad082951d72d257c68b5ade8fe73a3/pkg/controller/jobs/pod/pod_controller.go#L609

Since those pods need to be removed from the group list if they are finalized and deleted, but still in the cluster.